File: ftc_app-3.5/TeamCode/src/main/java/org/firstinspires/ftc/teamcode/libraries/vision/JewelPipeline.java
package org.firstinspires.ftc.teamcode.libraries.vision;

import org.corningrobotics.enderbots.endercv.OpenCVPipeline;
import org.opencv.core.Core;
import org.opencv.core.KeyPoint;
import org.opencv.core.Mat;
import org.opencv.core.MatOfKeyPoint;
import org.opencv.core.Point;
import org.opencv.core.Scalar;
import org.opencv.core.Size;
import org.opencv.features2d.FeatureDetector;
import org.opencv.imgproc.Imgproc;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

public class JewelPipeline extends OpenCVPipeline {

	//Outputs
	private Mat blurOutput = new Mat();
	private Mat hsvOutput = new Mat();
	private Mat hsvThreshold0Output = new Mat();
	private Mat cvErode0Output = new Mat();
	private Mat cvDilate0Output = new Mat();
	private MatOfKeyPoint findBlobs0Output = new MatOfKeyPoint();
	private Mat hsvThreshold1Output = new Mat();
	private Mat cvErode1Output = new Mat();
	private Mat cvDilate1Output = new Mat();
	private MatOfKeyPoint findBlobs1Output = new MatOfKeyPoint();

	private double lastKnownRedPosition = -1;
	private double lastKnownBluePosition = -1;

	static {
		System.loadLibrary(Core.NATIVE_LIBRARY_NAME);
	}

	@Override
	public Mat processFrame(Mat rgba, Mat gray) {
		double kernelSize = 2 * 10.0 + 1;
		Imgproc.blur(rgba, blurOutput, new Size(kernelSize, kernelSize));
		Imgproc.cvtColor(blurOutput, hsvOutput, Imgproc.COLOR_BGR2HSV);

		double[] blueHueThreshold = {0.0, 40.0};
		double[] blueSaturationThreshold = {150.0, 255.0};
		double[] blueValueThreshold = {0.0, 255.0};
		Core.inRange(hsvOutput, new Scalar(blueHueThreshold[0], blueSaturationThreshold[0], blueValueThreshold[0]), new Scalar(blueHueThreshold[1], blueSaturationThreshold[1], blueValueThreshold[1]), hsvThreshold0Output);
		Imgproc.erode(hsvThreshold0Output, cvErode0Output, new Mat(), new Point(-1, -1), 3,  Core.BORDER_CONSTANT, new Scalar(-1));
		Imgproc.dilate(cvErode0Output, cvDilate0Output, new Mat(), new Point(-1, -1), 6, Core.BORDER_CONSTANT, new Scalar(-1));

		double[] redHueThreshold1 = {60.0, 140.0};
		double[] redSaturationThreshold1 = {200.0, 255.0};
		double[] redValueThreshold1 = {0.0, 255.0};
		Core.inRange(hsvOutput, new Scalar(redHueThreshold1[0], redSaturationThreshold1[0], redValueThreshold1[0]), new Scalar(redHueThreshold1[1], redSaturationThreshold1[1], redValueThreshold1[1]), hsvThreshold1Output);
		Imgproc.erode(hsvThreshold1Output, cvErode1Output, new Mat(), new Point(-1, -1), 3,  Core.BORDER_CONSTANT, new Scalar(-1));
		Imgproc.dilate(cvErode1Output, cvDilate1Output, new Mat(), new Point(-1, -1), 6, Core.BORDER_CONSTANT, new Scalar(-1));

		return hsvThreshold1Output;
	}

	public enum JewelPositions {
		RED_BLUE,
		BLUE_RED,
		UNKNOWN
	}

	public double getRedPosition() {
		double[] findBlobs0Circularity = {0.7, 1.0};
		findBlobs(cvDilate0Output, 1000.0, findBlobs0Circularity, false, findBlobs0Output);

		KeyPoint[] redBlobs = findBlobs0Output.toArray();
		if(redBlobs.length > 0) {
			lastKnownRedPosition = redBlobs[0].pt.x;
			return lastKnownRedPosition;
		} else {
			return -1;
		}
	}

	public double getBluePosition() {
		double[] findBlobs1Circularity = {0.7, 1.0};
		findBlobs(cvDilate1Output, 1000.0, findBlobs1Circularity, false, findBlobs1Output);

		KeyPoint[] blueBlobs = findBlobs1Output.toArray();
		if(blueBlobs.length > 0) {
			lastKnownBluePosition = blueBlobs[0].pt.x;
			return lastKnownBluePosition;
		} else {
			return -1;
		}
	}

	public JewelPositions getJewelPositions() {
		final double bluePosition = getBluePosition();
		final double redPosition = getRedPosition();

		if(bluePosition != -1 && redPosition != -1) {
			if(bluePosition > redPosition) {
				return JewelPositions.BLUE_RED;
			} else {
				return JewelPositions.RED_BLUE;
			}
		} else {
			return JewelPositions.UNKNOWN;
		}
	}

	/**
	 * Detects groups of pixels in an image.
	 * @param input The image on which to perform the find blobs.
	 * @param minArea The minimum size of a blob that will be found
	 * @param circularity The minimum and maximum circularity of blobs that will be found
	 * @param darkBlobs The boolean that determines if light or dark blobs are found.
	 * @param blobList The output where the MatOfKeyPoint is stored.
	 */
	private void findBlobs(Mat input, double minArea, double[] circularity,
		Boolean darkBlobs, MatOfKeyPoint blobList) {
		FeatureDetector blobDet = FeatureDetector.create(FeatureDetector.SIMPLEBLOB);
		try {
			File tempFile = File.createTempFile("config", ".xml");

			String config = "<?xml version=\"1.0\"?>\n" +
					"<opencv_storage>\n" +
					"<thresholdStep>10.</thresholdStep>\n" +
					"<minThreshold>50.</minThreshold>\n" +
					"<maxThreshold>220.</maxThreshold>\n" +
					"<minRepeatability>2</minRepeatability>\n" +
					"<minDistBetweenBlobs>10.</minDistBetweenBlobs>\n" +
					"<filterByColor>1</filterByColor>\n" +
					"<blobColor>" +
					(darkBlobs ? 0 : 255) +
					"</blobColor>\n" +
					"<filterByArea>1</filterByArea>\n" +
					"<minArea>" +
					minArea +
					"</minArea>\n" +
					"<maxArea>" +
					Integer.MAX_VALUE +
					"</maxArea>\n" +
					"<filterByCircularity>1</filterByCircularity>\n" +
					"<minCircularity>" +
					circularity[0] +
					"</minCircularity>\n" +
					"<maxCircularity>" +
					circularity[1] +
					"</maxCircularity>\n" +
					"<filterByInertia>1</filterByInertia>\n" +
					"<minInertiaRatio>0.1</minInertiaRatio>\n" +
					"<maxInertiaRatio>" + Integer.MAX_VALUE + "</maxInertiaRatio>\n" +
					"<filterByConvexity>1</filterByConvexity>\n" +
					"<minConvexity>0.95</minConvexity>\n" +
					"<maxConvexity>" + Integer.MAX_VALUE + "</maxConvexity>\n" +
					"</opencv_storage>\n";

			FileWriter writer;
			writer = new FileWriter(tempFile, false);
			writer.write(config);
			writer.close();
			blobDet.read(tempFile.getPath());
		} catch (IOException e) {
			e.printStackTrace();
		}

		blobDet.detect(input, blobList);
	}
}

File: ftc_app-3.5/TeamCode/src/main/java/org/firstinspires/ftc/teamcode/libraries/vision/JewelPipeline.kt
package org.firstinspires.ftc.teamcode.libraries.vision

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.util.ElapsedTime
import org.corningrobotics.enderbots.endercv.OpenCVPipeline
import org.opencv.core.*
import org.opencv.features2d.FeatureDetector
import org.opencv.imgproc.Imgproc
import java.io.File
import java.io.FileWriter
import java.io.IOException

class JewelPipeline : OpenCVPipeline() {

    companion object {
        val TIME_OUT_SECONDS = 2.0

        init {
            System.loadLibrary(Core.NATIVE_LIBRARY_NAME)
        }
    }

    enum class JewelPositions {
        RED_BLUE,
        BLUE_RED,
        UNKNOWN
    }

    private val blurOutput = Mat()
    private val hsvOutput = Mat()

    private val bluePositiveThresholdOutput = Mat()
    private val blueNegativeThresholdOutput = Mat()
    private val blueThresholdOutput = Mat()
    private val blueErodeOutput = Mat()
    private val blueDilateOutput = Mat()
    private val findBlueBlobsOutput = MatOfKeyPoint()

    private val redThresholdOutput = Mat()
    private val redErodeOutput = Mat()
    private val redDilateOutput = Mat()
    private val findRedBlobsOutput = MatOfKeyPoint()

    private var lastKnownRedPosition = -1.0
    private var lastKnownBluePosition = -1.0

    val redPosition: Double
        get() {
            val circularity = doubleArrayOf(0.75, 1.0)
            findBlobs(redDilateOutput, 10000.0, circularity, false, findRedBlobsOutput)

            val redBlobs = findRedBlobsOutput.toArray()
            if (redBlobs.isNotEmpty()) {
                lastKnownRedPosition = redBlobs[0].pt.x

            }

            return lastKnownRedPosition
        }

    val bluePosition: Double
        get() {
            val circularity = doubleArrayOf(0.75, 1.0)
            findBlobs(blueDilateOutput, 10000.0, circularity, false, findBlueBlobsOutput)

            val blueBlobs = findBlueBlobsOutput.toArray()
            if (blueBlobs.isNotEmpty()) {
                lastKnownBluePosition = blueBlobs[0].pt.x
            }

            return lastKnownBluePosition
        }

    private val jewelPositions: JewelPositions
        get() {
            val bluePosition = bluePosition
            val redPosition = redPosition

            return if (bluePosition != -1.0 && redPosition != -1.0) {
                if (bluePosition < redPosition) {
                    JewelPositions.BLUE_RED
                } else {
                    JewelPositions.RED_BLUE
                }
            } else {
                JewelPositions.UNKNOWN
            }
        }

    override fun processFrame(rgba: Mat, gray: Mat): Mat {
        val kernelSize = 2 * 10.0 + 1
        Imgproc.blur(rgba, blurOutput, Size(kernelSize, kernelSize))

        Imgproc.cvtColor(blurOutput, hsvOutput, Imgproc.COLOR_BGR2HSV)

        val bluePositiveHueThreshold = doubleArrayOf(0.0, 40.0)
        val blueNegativeHueThreshold = doubleArrayOf(150.0, 180.0)
        val blueSaturationThreshold = doubleArrayOf(150.0, 255.0)
        val blueValueThreshold = doubleArrayOf(100.0, 255.0)
        Core.inRange(hsvOutput, Scalar(bluePositiveHueThreshold[0], blueSaturationThreshold[0], blueValueThreshold[0]), Scalar(bluePositiveHueThreshold[1], blueSaturationThreshold[1], blueValueThreshold[1]), bluePositiveThresholdOutput)
        Core.inRange(hsvOutput, Scalar(blueNegativeHueThreshold[0], blueSaturationThreshold[0], blueValueThreshold[0]), Scalar(blueNegativeHueThreshold[1], blueSaturationThreshold[1], blueValueThreshold[1]), blueNegativeThresholdOutput)
        Core.bitwise_or(bluePositiveThresholdOutput, blueNegativeThresholdOutput, blueThresholdOutput)
        Imgproc.erode(blueThresholdOutput, blueErodeOutput, Mat(), Point(-1.0, -1.0), 3, Core.BORDER_CONSTANT, Scalar(-1.0))
        Imgproc.dilate(blueErodeOutput, blueDilateOutput, Mat(), Point(-1.0, -1.0), 15, Core.BORDER_CONSTANT, Scalar(-1.0))

        val redHueThreshold = doubleArrayOf(110.0, 140.0)
        val redSaturationThreshold = doubleArrayOf(230.0, 255.0)
        val redValueThreshold = doubleArrayOf(100.0, 255.0)
        Core.inRange(hsvOutput, Scalar(redHueThreshold[0], redSaturationThreshold[0], redValueThreshold[0]), Scalar(redHueThreshold[1], redSaturationThreshold[1], redValueThreshold[1]), redThresholdOutput)
        Imgproc.erode(redThresholdOutput, redErodeOutput, Mat(), Point(-1.0, -1.0), 3, Core.BORDER_CONSTANT, Scalar(-1.0))
        Imgproc.dilate(redErodeOutput, redDilateOutput, Mat(), Point(-1.0, -1.0), 15, Core.BORDER_CONSTANT, Scalar(-1.0))

        return redDilateOutput
    }

    /**
     * Waits for the jewels to be identified until 5 seconds elapse.
     * @param elapsedTime The elapsed time instance containing the start time of the opmode.
     * @param linearOpMode The instance of LinearOpMode that is running.
     */
    fun waitForJewelIdentification(elapsedTime: ElapsedTime, linearOpMode: LinearOpMode): JewelPositions {
        while(jewelPositions == JewelPositions.UNKNOWN && elapsedTime.seconds() < TIME_OUT_SECONDS) {
            linearOpMode.telemetry.addData("Blue", bluePosition)
            linearOpMode.telemetry.addData("Red", redPosition)
            linearOpMode.telemetry.update()
            linearOpMode.sleep(10)
        }

        return jewelPositions
    }

    private fun findBlobs(input: Mat, minArea: Double, circularity: DoubleArray, darkBlobs: Boolean, blobList: MatOfKeyPoint) {
        val blobDet = FeatureDetector.create(FeatureDetector.SIMPLEBLOB)
        try {
            val tempFile = File.createTempFile("config", ".xml")

            val config = "<?xml version=\"1.0\"?>\n" +
                    "<opencv_storage>\n" +
                    "<thresholdStep>10.</thresholdStep>\n" +
                    "<minThreshold>50.</minThreshold>\n" +
                    "<maxThreshold>220.</maxThreshold>\n" +
                    "<minRepeatability>2</minRepeatability>\n" +
                    "<minDistBetweenBlobs>10.</minDistBetweenBlobs>\n" +
                    "<filterByColor>1</filterByColor>\n" +
                    "<blobColor>" +
                    (if (darkBlobs) 0 else 255) +
                    "</blobColor>\n" +
                    "<filterByArea>1</filterByArea>\n" +
                    "<minArea>" +
                    minArea +
                    "</minArea>\n" +
                    "<maxArea>" +
                    Integer.MAX_VALUE +
                    "</maxArea>\n" +
                    "<filterByCircularity>1</filterByCircularity>\n" +
                    "<minCircularity>" +
                    circularity[0] +
                    "</minCircularity>\n" +
                    "<maxCircularity>" +
                    circularity[1] +
                    "</maxCircularity>\n" +
                    "<filterByInertia>1</filterByInertia>\n" +
                    "<minInertiaRatio>0.1</minInertiaRatio>\n" +
                    "<maxInertiaRatio>" + Integer.MAX_VALUE + "</maxInertiaRatio>\n" +
                    "<filterByConvexity>1</filterByConvexity>\n" +
                    "<minConvexity>0.95</minConvexity>\n" +
                    "<maxConvexity>" + Integer.MAX_VALUE + "</maxConvexity>\n" +
                    "</opencv_storage>\n"

            val writer: FileWriter
            writer = FileWriter(tempFile, false)
            writer.write(config)
            writer.close()
            blobDet.read(tempFile.path)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        blobDet.detect(input, blobList)
    }
}